src/Algebra/Graph/Algorithm/TopSort/Ord.hs

```diff
@@ -0,0 +1,92 @@
+module Algebra.Graph.Algorithm.TopSort.Ord where
+
+import           Algebra.Graph
+import           Control.Monad.State.Lazy
+import           Data.Map.Strict          (Map)
+import qualified Data.Map.Strict          as Map
+import           Data.Set                 (Set)
+import qualified Data.Set                 as Set
+
+-- | O(s * log n)
+--
+-- Calculates topological number for each vertex.
+--
+-- >>> g = (1 * 2) + (2 * 3)
+-- >>> g_final = overlay (edge (0, 1) (1, 2)) (edge (1, 2) (2, 3))
+-- >>> topSort g == Just g_final
+-- True
+--
+-- If graph contains cycle, returns Nothing.
+--
+-- >>> g = (1 * 2) + (2 * 3) + (3 * 1)
+-- >>> topSort g
+-- Nothing
+topSort :: Ord a => Graph a -> Maybe (Graph (Int, a))
+topSort g = if isAcyclic g then Just $ evalState (traverseGraph g) s
+                                else Nothing
+  where
+    s = TopSortState 0 Map.empty
+
+data TopSortState a = TopSortState
+  { nextIndex     :: Int
+  , vertexIndices :: Map a Int
+  }
+
+traverseGraph :: Ord a => Graph a -> State (TopSortState a) (Graph (Int, a))
+traverseGraph Empty = return Empty
+traverseGraph (Vertex v) = do
+  TopSortState n_i vs <- get
+  case Map.lookup v vs of
+    Just i -> do
+      return $ Vertex (i, v)
+    Nothing -> do
+      put $ TopSortState (n_i + 1) (Map.insert v n_i vs)
+      return $ Vertex (n_i, v)
+traverseGraph (Overlay l r) = do
+  left <- traverseGraph l
+  right <- traverseGraph r
```

Comment:
Here left and right subgraphs should be traversed sort of "in parallel", meaning that we need to reset `nextIndex` before going into the right subgraph:

```haskell
do
  savedNextIndex <- gets nextIndex
  left <- traverseGraph l
  leftNextIndex <- gets nextIndex
  modify (\s -> s { nextIndex = savedNextIndex })
  right <- traverseGraph r
  rightNextIndex <- gets nextIndex
  modify (\s -> s { nextIndex = max leftNextIndex rightNextIndex })
  return (Overlay left right)
```